test/puppetlabs/puppetdb/integration/smoke.clj

```diff
@@ -27,3 +28,94 @@
+     {:name "deferred" :code "Deferred('join', [[1, 2, 3], ':'])" :expected "Deferred({'name' => 'join', 'arguments' => [[1, 2, 3], ':']})"}
+     {:name "sensitive_deferred" :code "Sensitive(Deferred('join', [[1, 2, 3], ':']))" :expected nil}
+     {:name "sensitive" :code "Sensitive('password')" :expected nil}
+     {:name "timestamp" :code "Timestamp('2012-10-10')" :expected "2012-10-10T00:00:00.000000000 UTC"}
+     {:name "hash_and_array" :code "{'a' => [1, 2, 3], 'b' => 'hello'}" :expected {:a [1, 2, 3] :b "hello"} }
+     {:name "special_key" :code "{ ['special', 'key'] => 10 }" :expected {(keyword "[\"special\", \"key\"]") 10}}
+     {:name "hash_with_sensitive_key" :code "{Sensitive(hush) => 42 }"
+      :valid? (fn [test-data]
+                (let [test-keys (keys test-data)]
+                  (and (= (count test-keys) 1)
+                       (re-matches #"#<Sensitive \[value redacted\]:[0-9]+>" (name (first test-keys)))
+                       (= 42 ((first test-keys) test-data)))))}
+     {:name "hash_with_sensitive_val" :code "{ x => Sensitive(hush) }"
+      :valid? (fn [test-data]
+                (and (= (count (keys test-data)) 1)
+                     (re-matches #"#<Sensitive \[value redacted\]:[0-9]+>" (:x test-data))))}
+     {:name "hash_ptype_key" :code "{'__ptype' => 10}" :expected {(keyword "reserved key: __ptype") 10}}
+     {:name "binary_value" :code "Binary(\"hello\", \"%s\")" :expected "aGVsbG8="}
+     {:name "a_type" :code "Integer[0,100]" :expected "Integer[0, 100]"}
+
+     ; Car is defined in the manifest generated in the test below
+     ; {:name "user_defined_type" :code "Car" :expected "Car"}
+     ; {:name "user_defined_object" :code "Car(abc123)" :expected "Car({'regnbr' => 'abc123'})"}
+     ])
+
+  (deftest ^:integration rich-data-agent-run
+    (with-open [pg (int/setup-postgres)
+                pdb (int/run-puppetdb pg {})
+                ps (int/run-puppet-server [pdb] {:agent {:rich-data true}})]
+
+      (testing "Agent run succeeds"
+        (let [manifest (str "type Car = Object[attributes => {regnbr => String}];\n"
+                            (apply str (for [{:keys [name code]} rich-data-tests]
+                                         (str "notify { '" name "':"
+                                              "  message => " code
+                                              "}"))))
+              {:keys [out]} (int/run-puppet-as "my-agent" ps pdb manifest {:rich-data true})]
+          (is (re-find #"a-z" out))))
+
+      (testing "Agent run data can be queried"
+        (are [query result] (= result (int/pql-query pdb query))
+          "nodes[certname] {}" [{:certname "my-agent"}]
+          "catalogs[certname, environment] {}" [{:certname "my-agent", :environment "production"}]
+          "factsets[certname, environment] {}" [{:certname "my-agent", :environment "production"}]
+          "reports[certname, environment] {}" [{:certname "my-agent", :environment "production"}]))
```

Comment:
If you prefer this, no worries, but these days I'm tempted to use "is" instead of "are" even if it's a little chattier so that the line numbers will be right when something goes wrong.

Reply:
That one is straight copy-pasta from the above test, so I'm tempted to leave it unless you want to change both

Reply:
Not critical either way.